contracts/marketplace/lib.rs

```diff
@@ -0,0 +1,323 @@
+        /// to `min_price()` over this period. The auction doesn't end after the period elapses -
+        /// the ticket remains available for purchase at `min_price()`.
+        #[ink(message)]
+        pub fn auction_length(&self) -> BlockNumber {
+            self.auction_length
+        }
+
+        /// The block at which the auction of the current ticket started.
+        #[ink(message)]
+        pub fn current_start_block(&self) -> BlockNumber {
+            self.current_start_block
+        }
+
+        /// The price the contract would charge when buying at the current block.
+        #[ink(message)]
+        pub fn price(&self) -> Balance {
+            self.current_price()
+        }
+
+        /// The average price over all sales the contract made.
+        #[ink(message)]
+        pub fn average_price(&self) -> Balance {
+            self.total_proceeds.saturating_div(self.tickets_sold)
+        }
+
+        /// The multiplier applied to the average price after each sale.
+        ///
+        /// The contract tracks the average price of all sold tickets and starts off each new
+        /// auction at `price() = average_price() * sale_multiplier()`.
+        #[ink(message)]
+        pub fn sale_multiplier(&self) -> Balance {
+            self.sale_multiplier
+        }
+
+        /// Number of tickets available for sale.
+        ///
+        /// The tickets will be auctioned off one by one.
+        #[ink(message)]
+        pub fn available_tickets(&self) -> Result<Balance, Error> {
+            self.ticket_balance()
+        }
+
+        /// The minimal price the contract allows.
+        #[ink(message)]
+        pub fn min_price(&self) -> Balance {
+            self.min_price
+        }
+
+        /// Update the minimal price.
+        #[ink(message)]
+        pub fn set_min_price(&mut self, value: Balance) -> Result<(), Error> {
+            Self::ensure_role(self.admin())?;
+
+            self.min_price = value;
+
+            Ok(())
+        }
+
+        /// Address of the reward token contract this contract will accept as payment.
+        #[ink(message)]
+        pub fn reward_token(&self) -> AccountId {
+            self.reward_token
+        }
+
+        /// Address of the ticket token contract this contract will auction off.
+        #[ink(message)]
+        pub fn ticket_token(&self) -> AccountId {
+            self.ticket_token
+        }
+
+        /// Buy one ticket at the current_price.
+        ///
+        /// The caller should make an approval for at least `price()` reward tokens to make sure the
+        /// call will succeed. The caller can specify a `max_price` - the call will fail if the
+        /// current price is greater than that.
+        #[ink(message)]
+        pub fn buy(&mut self, max_price: Option<Balance>) -> Result<(), Error> {
+            if self.ticket_balance()? <= 0 {
+                return Err(Error::MarketplaceEmpty);
```

Comment:
Maybe we should emit an event when the marketplace gets emptied? i.e. after the last ticket is sold. What do you think?

Reply:
It's easy enough to do, although I don't have a good feel yet of what types of things should generally be announced by events, so can't really say if it's a good idea or not.

Reply:
Otherwise, how would an observer of the chain find out that the system has run out of the tickets? One would have to call `buy()` to fail with an error. That has at least two cons compared with the event:
1. Someone has to call `buy()` and pay for the txn
2. The fact marketplace is empty is returned as an error which (to me) is coincidental

I think event here (and in general) would be more explicit and visible to any observer of the contract (since that's how events work, they're "tied" to a contract) rather than watching *all* txns, filtering down to ones that call the contract and checking the result of the call.

Reply:
You can use `psp22::balanceOf` to check the balance of tickets available to the market. Secondly, the market exposes `ticketsAvailable`, which is basically just a wrapper around `balanceOf`.

Reply:
Right, in that case I might be enough.